core/src/main/java/jenkins/model/LoginPageDecorator.java

```diff
@@ -0,0 +1,45 @@
+package jenkins.model;
+
+import hudson.DescriptorExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.Describable;
+import hudson.model.Descriptor;
+/**
+ * Participates in the rendering of the login page
+ *
+ * <p>
+ * This class provides a few hooks to augment the HTML of the login page.
+ *
+ * @since TODO
+ */
+public class LoginPageDecorator extends Descriptor<LoginPageDecorator> implements ExtensionPoint, Describable<LoginPageDecorator> {
+
+    protected LoginPageDecorator()  {
+        super(self());
+    }
+
+    @Override
+    public final Descriptor<LoginPageDecorator> getDescriptor() {
+        return this;
+    }
+    /**
+     * Obtains the URL of this object, excluding the context path.
+     *
+     * <p>
+     * Every {@link LoginPageDecorator} is bound to URL via {@link Jenkins#getDescriptor()}.
+     * This method returns such an URL.
+     */
+    public final String getUrl() {
+        return "descriptor/"+clazz.getName();
+    }
+
+    /**
+     * The first found LoginDecarator, there can only be one.
```

Comment:
It is unclear how this is supposed to work in a world with countless plugins. Install a plugin that has `POSITIVE_INFINITY` and everything else loses out?

Reply:
The login provides a mean to override some decorators , since we are treating stuff like logo and welcome message they need to be unique. If a plugin provides a custom implementation for the login page then this needs to be unique and yes everyone else loses out, that is the whole point about it.